### PR TITLE
Cleanup feature_test_macros/test.cpp

### LINKED ISSUE #600: VSO_0157762_feature_test_macros/test.cpp: Needs comment/error cleanup
I wrote this feature-test macro test, which has proven to be difficult to maintain:

https://github.com/microsoft/STL/blob/285187b7b24be4bebfa271fbd0b1eacd995d18f4/tests/std/tests/VSO_0157762_feature_test_macros/test.cpp#L35-L42

First, `#error BOOM` isn't helpful. Instead, we should say something like `#error Expected __cpp_constexpr to be defined.` With VSCode multi-line regex search-and-replace, we should be able to make such changes systematically.

Second, nobody notices the comments indicating various sections, like `// Always defined to varying values (C++14 versus C++17-and-newer mode).` Nor do these really add value. Each preprocessor chunk self-documents what it expects, and we've accumulated a number of exceptions (as C1XX, Clang, and EDG implement features on different schedules). We've damaged this organization repeatedly during maintenance and it provides no value, so it should be removed. Instead, we should simply sort the preprocessor chunks lexicographically.

Third, this unusually covers both the compiler and the STL (for the historical reason that I implemented almost all of the feature-test macros in the compiler). We should consider splitting up this test, handing off the compiler part to the compiler team, and making the library part simply trust that `_HAS_CXX17` and `_HAS_CXX20` have been properly defined, instead of replicating that machinery with:

https://github.com/microsoft/STL/blob/285187b7b24be4bebfa271fbd0b1eacd995d18f4/tests/std/tests/VSO_0157762_feature_test_macros/test.cpp#L10-L32

### PR DESCRIPTION
**Pull request by SuperWig**
Description
===========

Fixes #600, _Hopefully_ I haven't messed this up. Reordering was...an experience,

* Used the forms "Expected `foo` to (NOT) be defined." / "Expected `foo` to equal `value`." To replace `BOOM`
* Removed `CXX??_MODE` and replaced with `_HAS_CXX??`
* Swapped the order of `#if !defined(clang) || _HAS_CXX17` for a better sorted order
* Made "Attributes" comment uppercase to follow the library section comment (~~and new language section comment~~ realised it already existed I just accidentally deleted it in the first commit)
* Reordered the tests removing `// Always defined...` sections. Tests are now ordered lexicographically grouped by Attributes, Language, then Library.

Checklist
=========

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [ ] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [ ] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [ ] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [ ] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

**Comment by SuperWig** [comment]
Looks like I've deleted some, Should be ~1805 lines not 1780. Let me try again...

File: tests/std/tests/VSO_0157762_feature_test_macros/test.cpp [hunk 1]
```diff
@@ -472,1396 +438,1366 @@ STATIC_ASSERT(__cpp_structured_bindings == 201606L);
...
+#else
+STATIC_ASSERT(__cpp_conditional_explicit == 201806L);
+#endif
+#else
+#ifdef __cpp_conditional_explicit
+#error Expected __cpp_conditional_explicit to NOT be defined.
 #endif
 #endif
 
 // C++98, /EHs[-]
```
Comment (SuperWig) [inline tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: I sorted the sections using msort so these got sorted by their comments (except `//C++98...` which I manually adjusted). Should they be kept that way or sorted using the condition?

File: tests/std/tests/VSO_0157762_feature_test_macros/test.cpp [hunk 2]
```diff
@@ -5,465 +5,431 @@
 
 int main() {} // COMPILE-ONLY
 
-// COMPILER FEATURE-TEST MACROS
+// ATTRIBUTE FEATURE-TEST MACROS.
```
Comment (CaseyCarter) [inline tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: "Attribute feature-test macros" isn't a sentence, so it should not end with a period:
```suggestion
// ATTRIBUTE FEATURE-TEST MACROS
```

File: tests/std/tests/VSO_0157762_feature_test_macros/test.cpp [hunk 3]
```diff
@@ -5,465 +5,431 @@
...
-#define CXX20_MODE 1
-#elif _MSVC_LANG > 201402L
-#define CXX17_MODE 1
-#define CXX20_MODE 0
+#ifdef __has_cpp_attribute
+// Good.
 #else
-#define CXX17_MODE 0
-#define CXX20_MODE 0
+#error Expected __has_cpp_attribute to be defined.
```
Comment (CaseyCarter) [inline tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: In general, I think I'd prefer the more direct "`meow` is not `woof`" to "Expected `meow` to be `woof`" with the implied "...which it is not". (If you don't want to rewrite all of these messages a dozen times, I suggest giving other reviewers a chance to check in before making any sweeping changes.)

File: tests/std/tests/VSO_0157762_feature_test_macros/test.cpp [hunk 4]
```diff
@@ -5,465 +5,431 @@
...
-#if __cpp_constexpr != 201603L
-#error BOOM
-#else
-STATIC_ASSERT(__cpp_constexpr == 201603L);
+#if __has_cpp_attribute(noreturn) != 200809L
+#error Expected has_cpp_attribute(noreturn) to equal 200809L.
+#endif
+
+#if defined(__clang__) || defined(__EDG__) // clang and EDG don't yet implement P1771R1
+#if __has_cpp_attribute(nodiscard) != 201603L
```
Comment (CaseyCarter) [inline tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: "nodiscard" < "noreturn" [My expectation is that within the attribute/core/library sections the tests are ordered lexicographically by the name of the attribute/feature-test macro to reflect the order in which the standard defines them.]

File: tests/std/tests/VSO_0157762_feature_test_macros/test.cpp [hunk 5]
```diff
@@ -5,465 +5,431 @@
...
-#ifndef __cpp_inheriting_constructors
-#error BOOM
-#elif (CXX17_MODE || defined(__clang__)) && !defined(__EDG__) // Clang implemented this C++17 feature unconditionally.
-                                                              // TRANSITION, VSO-610203
-#if __cpp_inheriting_constructors != 201511L
-#error BOOM
-#else
-STATIC_ASSERT(__cpp_inheriting_constructors == 201511L);
+#if _HAS_CXX17 || defined(__clang__) || defined(__EDG__) // Clang and EDG provide this in C++14 mode.
+#if __has_cpp_attribute(fallthrough) != 201603L
```
Comment (CaseyCarter) [inline tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: Ditto "fallthrough isn't in the order I expect from the issue discussion." (I'll stop commenting on ordering now - there are a lot of these that don't match my expectations.)

File: tests/std/tests/VSO_0157762_feature_test_macros/test.cpp [hunk 6]
```diff
@@ -5,465 +5,431 @@
...
 #endif
 #endif
 
+#if __has_cpp_attribute(noreturn) != 200809L
+#error Expected has_cpp_attribute(noreturn) to equal 200809L.
+#endif
 
-// Always defined to specific values.
+
+// COMPILER FEATURE-TEST MACROS
```
Comment (CaseyCarter) [inline tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: I think I'd prefer "core" or "core language" to "compiler" here to reflect which part of the Standard defines these macros.

File: tests/std/tests/VSO_0157762_feature_test_macros/test.cpp [hunk 7]
```diff
@@ -5,465 +5,431 @@
...
 
-#ifdef _MSVC_LANG
-#if _MSVC_LANG > 201703L
-#define CXX17_MODE 1
-#define CXX20_MODE 1
-#elif _MSVC_LANG > 201402L
-#define CXX17_MODE 1
-#define CXX20_MODE 0
+#ifdef __has_cpp_attribute
+// Good.
```
Comment (CaseyCarter) [inline tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: Ditto (pre-existing) "this isn't a sentence or a floating-point literal so it shouldn't have a period at the end"
```suggestion
// Good
```

File: tests/std/tests/VSO_0157762_feature_test_macros/test.cpp [hunk 8]
```diff
@@ -5,465 +5,431 @@
...
 #else
-#define CXX17_MODE 0
-#define CXX20_MODE 0
+#error Expected __has_cpp_attribute to be defined.
 #endif
+#if defined(__has_cpp_attribute)
+// Good.
```
Comment (CaseyCarter) [inline tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: Ditto (pre-existing) "this isn't a sentence or a floating-point literal so it shouldn't have a period at the end"
```suggestion
// Good
```

File: tests/std/tests/VSO_0157762_feature_test_macros/test.cpp (hunk 1 above)
Comment (CaseyCarter) [reply tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: IMO, they should be sorted lexicographically by the name under test.

File: tests/std/tests/VSO_0157762_feature_test_macros/test.cpp (hunk 6 above)
Comment (SuperWig) [reply tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: Yeah I had "language" before realising I deleted the existing comment marking the section. Going with "core language"

File: tests/std/tests/VSO_0157762_feature_test_macros/test.cpp (hunk 3 above)
Comment (SuperWig) [reply tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: Well there's also inverse cases so would you want "`meow` is `woof`" too? 
In the current form I capitalised NOT as I thought it could easily be missed. Then again with the shorter form it'll be harder to miss? 🤷‍♂️

File: tests/std/tests/VSO_0157762_feature_test_macros/test.cpp (hunk 4 above)
Comment (SuperWig) [reply tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: The tests were also originally ordered by standard mode which is how I have kept them. Should that be kept at all or just completely lexicographically by test name?

Comment (cbezault) [reply tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: The issue stated that lexicographically was the proposed solution since there doesn't seem to be much value keeping them ordered by standard mode.
I would actually propose that we sort them lexicographically now but in the future we move to just appending new tests to the end of the file so in the future they are chronologically ordered by appearance in our library or in one of the supported compilers.

Comment (SuperWig) [reply tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: Great! Because that was already what I was working on and now have just pushed.

Comment (cbezault) [reply tests/std/tests/VSO_0157762_feature_test_macros/test.cpp]: After talking to @StephanTLavavej I take back my appending comment. Let's keep this file sorted in the future as well.
